### PR TITLE
Resolve #468

### LINKED ISSUE #468: Incorrect handling of scale in `Loss.grad`
There is a bug in `Loss.grad` handling of the `scale` attribute, but only when it's set via scalar multiplication:

```python
import jax
from scico.loss import SquaredL2Loss
from scico.functional import L2Norm
import scico.numpy as snp

f = SquaredL2Loss(y=snp.zeros((4,)))
g = SquaredL2Loss(y=snp.zeros((4,)), scale=5)
h = 10 * f

# __call__ is correct
f(snp.ones((4,)))
>> Array(2., dtype=float32)
g(snp.ones((4,)))
>> Array(20., dtype=float32)
h(snp.ones((4,)))
>> Array(20., dtype=float32)

# __grad__ is broken
f.grad(snp.ones((4,)))
>> Array([1., 1., 1., 1.], dtype=float32)
g.grad(snp.ones((4,)))
>> Array([10., 10., 10., 10.], dtype=float32)
h.grad(snp.ones((4,)))
>> Array([1., 1., 1., 1.], dtype=float32
```

The same bug is not present in `Functional.grad`:
```python
f = L2Norm()
g = 10 * f

f.grad(snp.ones((4,)))
>> Array([0.5, 0.5, 0.5, 0.5], dtype=float32)
g.grad(snp.ones((4,)))
>> Array([5., 5., 5., 5.], dtype=float32)
```

### PR DESCRIPTION
**Pull request by bwohlberg**
Resolve #468.

Also:
* Improve scaling of `ScaledFunctional`, so that e.g. ` 5.0 * (2.0 * functional.L2Norm())` is a `ScaledFunctional` with a scale of 10 of an `L2Norm` rather than a `ScaledFunctional` with a scale of 5 of a `ScaledFunctional`.
* Some cleaning up and code re-organization.
* Update change summary `CHANGES.rst`.

### CODE REVIEW

**Review by Michael-T-McCann — CHANGES_REQUESTED** [review]

File: scico/functional/_functional.py [hunk 1]
```diff
@@ -38,15 +43,15 @@ def __init__(self):
     def __repr__(self):
         return f"""{type(self)} (has_eval = {self.has_eval}, has_prox = {self.has_prox})"""
 
-    def __mul__(self, other):
+    def __mul__(self, other: Union[float, int]) -> ScaledFunctional:
         if snp.isscalar(other) or isinstance(other, jax.core.Tracer):
```
[the remark this answers is not in the document]
Comment (bwohlberg) [reply scico/functional/_functional.py]: Good point. Done.

**Review by Michael-T-McCann — APPROVED** [review]
